### PR TITLE
Spring Boot Webflux Netty Tests

### PR DESCRIPTION
**Pull request by gary-huang**
Tests out tracing spring boot webflux 2 using Netty as the underlying server.

### CODE REVIEW

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code() == 200
+    expectedResponseBodyStr == response.body().string()
+    assertTraces(TEST_WRITER, 1) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: I'm interested in seeing the `resource.name` here too.

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            defaultTags()
+          }
+        }
+      }
+    }
+  }
+
+  def "Flux GET test"() {
+    setup:
+    def count = 10
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: stylistically, I like to define arbitrary test parameters in a `where` clause.  No big deal though.

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    when:
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code == 200
+    assertTraces(TEST_WRITER, 2) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: Add `resource.name`

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    when:
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code() == 500
+    assertTraces(TEST_WRITER, 1) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: Add `resource.name`

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            "$DDTags.SPAN_TYPE" DDSpanTypes.HTTP_SERVER
+            "$Tags.PEER_HOSTNAME.key" "localhost"
+            "$Tags.PEER_PORT.key" Integer
+            "$Tags.HTTP_METHOD.key" "POST"
+            "$Tags.HTTP_STATUS.key" 500
+            "$Tags.HTTP_URL.key" url
+            "error" true
+            // errorTags(NullPointerException, String)
+            // ideally this should be here with the stacktrace, but for now this doesn't seem
+            // like a low hanging fruit/requires further instrumentation at spring webflux level
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: I say we add that instrumentation now, along with proper `resource.name` instrumentation, similar to `spring-web`'s instrumentation.

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code() == 202
+    response.body().string() == echoString
+    assertTraces(TEST_WRITER, 1) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: Add `resource.name`

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+    when:
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code == 404
+    assertTraces(TEST_WRITER, 1) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: Add `resource.name`

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    def response = client.newCall(request).execute()
+
+    then:
+    response.code == 200
+    response.body().string() == expectedResponseBody
+    assertTraces(TEST_WRITER, 1) {
+      trace(0, 1) {
+        span(0) {
+          operationName "netty.request"
+          spanType DDSpanTypes.HTTP_SERVER
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: Add `resource.name`

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+  static final okhttp3.MediaType PLAIN_TYPE = okhttp3.MediaType.parse("text/plain; charset=utf-8")
+
+  @Shared
+  SpringApplication app
+
+  @Shared
+  ConfigurableApplicationContext context
+
+  @Shared
+  String randomServerPort = -1
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: rename to just `port` and don't assign the `-1`.  It gets initialized in the setup anyway.

File: dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+  @Shared
+  String baseUrl = "http://localhost:"
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/src/test/groovy/SpringWebfluxTest.groovy]: remove this line.  just hardcode it in the tests.  instead do something like "http://localhost:$port/somepath" anywhere you're using this.

File: dd-java-agent/instrumentation/netty-4.1/netty-4.1.gradle [hunk 11]
```diff
@@ -36,6 +36,12 @@ dependencies {
   testCompile group: 'org.eclipse.jetty', name: 'jetty-server', version: '8.2.0.v20160908'
   testCompile group: 'org.asynchttpclient', name: 'async-http-client', version: '2.0.31'
   // async-http-client:2.0.32+ would require netty:4.1.9.Final
+
+  testCompile group: 'org.springframework.boot', name: 'spring-boot-starter-webflux', version: '2.0.0.RELEASE'
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/netty-4.1/netty-4.1.gradle]: I'd prefer moving these tests into a `spring-webflux` instrumentation project to go along with any additional instrumentation we add.  That project should then have a test dependency on netty's instrumentation.

Comment (gary-huang) [reply dd-java-agent/instrumentation/netty-4.1/netty-4.1.gradle]: Yes, I agree, now that we are instrumenting `spring-webflux` I will move these tests to that project. :)

File: dd-java-agent/instrumentation/spring-webflux/src/main/java/datadog/trace/instrumentation/springweb/ErrorHandlerFunctionInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
+package datadog.trace.instrumentation.springweb;
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/main/java/datadog/trace/instrumentation/springweb/ErrorHandlerFunctionInstrumentation.java]: Change package to `springwebflux`.

**Comment by gary-huang** [comment]
Closing in favor of #460

**Pull request closed without merging**